### PR TITLE
FLY-2520: P2SH-P2WSH derivation path for segwit powpeg federations

### PR DESCRIPTION
**Pull request by TravellerOnTheRun**
## What

Switches `PegInDerivation` step 4 to the P2SH-P2WSH wrapping, so commit-first deposit addresses match what the live bridge derives under a segwit (`P2SH_P2WSH_ERP_FEDERATION`) powpeg. Segwit is the only supported format: per review, the plain-P2SH path and the on-chain format inference it required were dropped. `PegInAddressRegistry` is unchanged apart from a comment — it derives through the same library call as before, and there is no `IPegInAddressRegistry` ABI or storage change.

## Why

Mainnet and testnet powpegs already use segwit federation wrapping. Plain-P2SH-only derivation issues addresses the bridge will not settle (`-304` VALUE_ZERO polarity when wrapping is wrong). This is the correctness fix for live federations (FLY-2520 / Stage 04 plan).

## Type of Change

- [ ] Bug fix (non-breaking change)
- [x] New feature (non-breaking change)
- [ ] Breaking change
- [ ] Refactor (no intended behavior change)
- [x] Security hardening
- [ ] Tests only
- [ ] Build/CI/Tooling
- [x] Documentation

## Affected Areas

- [ ] PegIn flow (`registerPegIn`, `callForUser`, quote validation)
- [ ] PegOut flow (`depositPegout`, `refundPegOut`, `refundUserPegOut`)
- [ ] Liquidity Provider lifecycle (register/update/status/resign/collateral/withdraw)
- [ ] Pause/operational controls
- [ ] Quote hashing/signature logic
- [x] Bitcoin tx / PMT / merkle validation
- [ ] Deployment or upgrade scripts / Makefile targets
- [ ] CI workflows (`ci`, `fuzz`, `formal`, `slither`, `codeql`)
- [x] Docs

Also: `PegInAddressRegistry` derivation / registration deposit matching (commit-first path). Quote-path `PegInContract.validatePegInDepositAddress` intentionally untouched. Base is `v3.0.0` (includes `resolvePegIn` / watchtower).

## Risk & Security Review

- [x] Access control and authorization paths reviewed
- [x] Reentrancy and external call paths reviewed
- [x] Storage layout / upgrade safety reviewed (if upgradeable code changed)
- [ ] Time/block/confirmation assumptions reviewed
- [x] BTC tx output/index assumptions reviewed (if pegout validation touched)
- [x] No secrets or sensitive values added

Notes: no registry storage or external-call change (the registry diff is a comment only); storage-layout inspect expected unchanged; ABI of `IPegInAddressRegistry` unchanged.

## Test Plan

List what you ran locally and the outcome:

- [x] `npm run lint:sol`
- [x] `npm run lint:ts`
- [x] `npm run compile`
- [x] `npm test`
- [ ] `npm run test:fuzz` (if logic/state transitions changed)
- [ ] `npm run test:invariant` (if invariant-sensitive logic changed)
- [ ] `npm run test:formal` (if formal-verification-sensitive logic changed)
- [ ] `npm run test:integration` (if integration paths changed)
- [ ] `npm run test:e2e` (if deployment/ownership flows changed)

Additional notes on test coverage, edge cases, and any tests intentionally skipped:

- `make test-unit` green: 627 tests, 0 failures.
- `forge test --match-path 'test/libraries/PegInDerivation.t.sol'` and `test/pegin-registry/*.t.sol` green — all fixtures re-pinned to the segwit form, plain-P2SH wrapping kept as the pitfall #2 negative pin.
- Storage layout / ABI checks: no diff on registry / interface.
- **Segwit-federation live regtest e2e: NOT complete.** Foundry pins + registry mocks are the merge gate; must be run as a precondition before the first testnet deployment after this change.

## Deployment & Ops Impact

- [x] No deployment impact
- [ ] Requires deployment
- [ ] Requires upgrade
- [ ] Env/config changes required (`.env`, network RPCs, verification, etc.)
- [x] Runbook/update instructions added to PR description

No deploy in this PR. Once deployed, every derived deposit address rotates (as it does on any federation change), and the registry must only ever run against a segwit powpeg — the constraint is documented in the `PegInDerivation` header. **Do not cut over to testnet until the segwit regtest e2e is completed.**

## Related Issues

- Jira: [FLY-2520](https://rsklabs.atlassian.net/browse/FLY-2520)
- GitHub Issue: #(if applicable)

## Reviewer Notes

- Focus on `PegInDerivation.flyoverScriptHash` / `witnessProgram` byte-matching the `P2SH_P2WSH_ERP_FEDERATION` branch of rskj `PegUtils.getFlyoverFederationOutputScript`.
- The tripwire against the quote path became a wrapping pin (`test_derivation_wrapping_matches_pegin_contract`): both paths now wrap identically and only differ in the derivation value they embed.
- **Known limitation:** a pre-segwit powpeg is deliberately unsupported and would derive unsettleable addresses; called out in the library header.
- **Known limitation:** live segwit-federation regtest settle e2e was not run for this PR; track as pre-testnet gate.
- Branch note: rebased onto `v3.0.0` as `fly-2520-v3` because force-push to `fly-2520` is blocked by repo rules (old `fly-2520` was cut from pre-rename `3.0.0`).

## Screenshots / Logs (if applicable)

N/A — contract/library change; CI / Foundry suite is the evidence for merge.

### CODE REVIEW

File: docs/pegin/federation-format-detection.md
Comment (Luisfc68) [inline docs/pegin/federation-format-detection.md:1]: As discussed in slack and refinement, I think we don't need the multiple format support. The rationale is that there are already multiple forks on top of the one that migrated to segwit and having this code implies maintainance and everything it involves

File: src/libraries/PegInDerivation.sol [hunk 1]
```diff
@@ -109,29 +130,139 @@ library PegInDerivation {
...
+            return FederationFormat.SegwitP2SHP2WSH;
+        }
+        if (federationScriptHash == plainCandidate) {
+            return FederationFormat.PlainP2SH;
+        }
+        revert UnrecognizedFederationFormat();
+    }
+
+    /// @notice Extracts the 20-byte script hash from a base58check-encoded P2SH address.
+    function _scriptHashFromBase58check(
```
Comment (Luisfc68) [inline src/libraries/PegInDerivation.sol]: When you remove the multiple format support I recommend to remove this too (since rn is only being used there), decoding can be done off chain so there's no need of spending the extra gas here

File: test/libraries/PegInDerivation.t.sol [hunk 2]
```diff
@@ -183,6 +217,204 @@ contract PegInDerivationTest is Test {
         );
     }
 
+    function test_Step4_SegwitScriptHash() public pure {
```
Comment (Luisfc68) [inline test/libraries/PegInDerivation.t.sol]: Similar to my cocern about referencing the walkthrough in the comments, I also don't recomment doing it in the tests, specially in the tests, remeber that this code is being audited publicly, so this will force us either to push misleading code/tests or to make public internal documents, I highly recommend avoiding references to the walkthrough in the code, and if is extremely necessary, do it only in the comments and clean up once the epic is finished

**Review by Luisfc68 — CHANGES_REQUESTED** [review]
The rest of my comments were already discussed in the refinement and also in the slack thread

File: docs/pegin/federation-format-detection.md
Comment (TravellerOnTheRun) [reply docs/pegin/federation-format-detection.md:1]: Agreed — multi-format support is gone. Derivation is now segwit-only (`P2SH_P2WSH_ERP_FEDERATION`), this doc is deleted, and the "only segwit powpegs are supported" constraint lives in the `PegInDerivation` header instead. The registry no longer reads `getFederationAddress()` at all, so the diff there is a comment only.

File: src/libraries/PegInDerivation.sol (hunk 1 above)
Comment (TravellerOnTheRun) [reply src/libraries/PegInDerivation.sol]: Removed. `_scriptHashFromBase58check`, the `Base58` import and both inference errors are gone, so there is no base58 decoding (and no assembly, which also clears the Slither finding) on-chain anymore.

File: test/libraries/PegInDerivation.t.sol (hunk 2 above)
Comment (TravellerOnTheRun) [reply test/libraries/PegInDerivation.t.sol]: Done. Dropped the walkthrough references from every file this PR touches (`PegInDerivation.sol`, `PegInAddressRegistry.sol` and this test file), along with the "PoC"/internal-run wording on the fixtures. Note there are still pre-existing walkthrough references in `IPegInAddressRegistry.sol`, `IPegInCommitFirst.sol` and `IFlyoverConfigurations.sol` — left untouched here to keep this diff scoped; happy to clean them up in a separate PR.

**Review by Luisfc68 — APPROVED** [review]
